### ISSUE TITLE
Example with props

### ISSUE THREAD

**Issue by stripuramallu3** [pmndrs/react-three-offscreen#8]
Could an example be provided where the Scene component has to receive props (for example for the mesh position)?

`// Scene.jsx (a self contained webgl app)
export default function App() {
  return (
    <mesh>
      <boxGeometry />
    </mesh>
  )
}`

**Comment by drcmda** [comment]
dom and canvas run into two different threads, it's up to you how you bridge that. broadcasts maybe, but there's no conventional way.

**Comment by stripuramallu3** [comment]
Thank you! Was hoping there was a better/conventional way

**Comment by sshmaxime** [comment]
It would be so beautiful to pass a `ref` to the worker canvas and manipulate it from the main thread. I guess that's not doable, but why ?

**Comment by drcmda** [comment]
if you guys have any ideas go ahead, would love to have a standard interface of some sort.

**Comment by mhmdjaw** [comment]
Does that also mean that `<Html>` from drei doesn't work in offscreen canvas? I tried it and it doesn't show any HTML elements.

**Comment by rob-myers** [comment]
To get this working, we can patch `@react-three/offscreen@0.0.8` as follows:

<details>

<summary>Show <i>patches/@react-three+offscreen+0.0.8.patch</i></summary>

```diff
diff --git a/node_modules/@react-three/offscreen/dist/Canvas.d.ts b/node_modules/@react-three/offscreen/dist/Canvas.d.ts
index 7084e67..e79c825 100644
--- a/node_modules/@react-three/offscreen/dist/Canvas.d.ts
+++ b/node_modules/@react-three/offscreen/dist/Canvas.d.ts
@@ -13,5 +13,6 @@ export interface CanvasProps extends Omit<RenderProps<HTMLCanvasElement>, 'size'
     eventSource?: HTMLElement | React.MutableRefObject<HTMLElement>;
     /** The event prefix that is cast into canvas pointer x/y events, default: "offset" */
     eventPrefix?: 'offset' | 'client' | 'page' | 'layer' | 'screen';
+    sceneProps?: Record<string, any>;
 }
 export declare function Canvas({ eventSource, worker, fallback, style, className, id, ...props }: CanvasProps): JSX.Element;
diff --git a/node_modules/@react-three/offscreen/dist/index.mjs b/node_modules/@react-three/offscreen/dist/index.mjs
index acf1918..9282957 100644
--- a/node_modules/@react-three/offscreen/dist/index.mjs
+++ b/node_modules/@react-three/offscreen/dist/index.mjs
@@ -86,6 +86,7 @@ function Canvas({
   style,
   className,
   id,
+  sceneProps = {},
   ...props
 }) {
   const [shouldFallback, setFallback] = React.useState(false);
@@ -111,7 +112,7 @@ function Canvas({
     worker.postMessage({
       type: 'init',
       payload: {
-        props,
+        props: { sceneProps, ...props },
         drawingSurface: offscreen,
         width: canvas.clientWidth,
         height: canvas.clientHeight,
@@ -181,9 +182,9 @@ function Canvas({
     if (!worker) return;
     worker.postMessage({
       type: 'props',
-      payload: props
+      payload: { sceneProps, ...props }
     });
-  }, [worker, props]);
+  }, [worker, props, sceneProps]);
   return shouldFallback ? /*#__PURE__*/React.createElement(Canvas$1, _extends({
     id: id,
     className: className,
@@ -203,7 +204,7 @@ function Canvas({
   });
 }
 
-function render(children) {
+function render(component) {
   extend(THREE);
   let root;
   let dpr = [1, 2];
@@ -218,6 +219,7 @@ function render(children) {
   const handleInit = payload => {
     const {
       props,
+      sceneProps,
       drawingSurface: canvas,
       width,
       top,
@@ -281,7 +283,7 @@ function render(children) {
       });
 
       // Render children once
-      root.render(children);
+      root.render(React.createElement(component, sceneProps));
     } catch (e) {
       postMessage({
         type: 'error',
@@ -317,14 +319,15 @@ function render(children) {
       stopPropagation() {}
     });
   };
-  const handleProps = payload => {
+  const handleProps = ({ sceneProps, ...payload }) => {
     if (!root) return;
     if (payload.dpr) dpr = payload.dpr;
     root.configure({
       size,
       dpr,
-      ...payload
+      ...payload,
     });
+    root.render(React.createElement(component, sceneProps));
   };
   const handlerMap = {
     resize: handleResize,
diff --git a/node_modules/@react-three/offscreen/dist/render.d.ts b/node_modules/@react-three/offscreen/dist/render.d.ts
index 2be3f9c..0cceb2b 100644
--- a/node_modules/@react-three/offscreen/dist/render.d.ts
+++ b/node_modules/@react-three/offscreen/dist/render.d.ts
@@ -1,2 +1,2 @@
 /// <reference types="react" />
-export declare function render(children: React.ReactNode): void;
+export declare function render<P extends {}>(component: React.FunctionComponent<P> | React.ComponentClass<P>): void;

```

</details>

Finally, restructure the project a bit:

```js
// create-worker.js
// separate file fixes HMR
export const worker = new Worker(new URL('./worker.jsx', import.meta.url), { type: 'module' });
```

```jsx
// App.jsx
import { worker } from "./create-worker";
// ...

  return (
      <Canvas
        fallback={<Scene testProp="hello, world! (fallback)" />}
        worker={worker}
        sceneProps={{ foo: "hello, world!" }}
      />
  );
```

```jsx
// worker.jsx
import { render } from '@react-three/offscreen';
import Scene from "./Scene";

// Send component instead of ReactElement
render(Scene);

**Comment by richi-coder** [comment]
> Does that also mean that `<Html>` from drei doesn't work in offscreen canvas? I tried it and it doesn't show any HTML elements.

How have you dealt with this so far? @mhmdjaw

**Comment by opeologist** [comment]
> To get this working, we can patch `@react-three/offscreen@0.0.8` as follows:
> 
> Show patches/@react-three+offscreen+0.0.8.patch
> Finally, restructure the project a bit:
> 
> // create-worker.js
> // separate file fixes HMR
> export const worker = new Worker(new URL('./worker.jsx', import.meta.url), { type: 'module' });
> // App.jsx
> import { worker } from "./create-worker";
> // ...
> 
>   return (
>       <Canvas
>         fallback={<Scene testProp="hello, world! (fallback)" />}
>         worker={worker}
>         sceneProps={{ foo: "hello, world!" }}
>       />
>   );
> // worker.jsx
> import { render } from '@react-three/offscreen';
> import Scene from "./Scene";
> 
> // Send component instead of ReactElement
> render(Scene);

@rob-myers apologies for digging up this old comment, but i noticed you said that putting the new Worker() declaration as an export in a separate file, which is supposed to fix hot module reloading, but i'm not seeing that work for me :( any time i update my scene that this package is rendering, i have to reload the page. was there some additional thing needed to be done besides putting it in a separate file? thanks!

**Comment by rob-myers** [comment]
@opeologist Not sure about your setup (or even about this project). You might be able to fix things using a basic trick:

```tsx
// App component
React.useEffect(() => {
  if (process.env.NODE_ENV !== "production") {
    // HMR this file onchange worker files
    import('./worker.jsx');
  }
}, []);
```

**Comment by opeologist** [comment]
> [@opeologist](https://github.com/opeologist) Not sure about your setup (or even about this project). You might be able to fix things using a basic trick:
> 
> // App component
> React.useEffect(() => {
>   if (process.env.NODE_ENV !== "production") {
>     // HMR this file onchange worker files
>     import('./worker.jsx');
>   }
> }, []);

hm, unfortunately that didn't seem to work... ideally i'd like to not rely on the fallback/standard canvas in order to get HMR working, so i can make sure the offscreencanvas tech works as expected during development, but it doesn't seem like there's a solution that works here.

my setup isn't complicated, it follows how you set up yours above, where:

- worker's exported from a separate file (src/workers/index.ts)
- worker file uses @react-three/offscreen's `render()`
- added the use effect you mentioned in my app component

is there a sample repo or codesandbox example that shows hot module reloading working as expected? i just don't want to have to refresh every time i make a change to the scene that offscreencanvas is rendering 😢

**Comment by rob-myers** [comment]
Hey @opeologist just had a go, but couldn't get the basics working.

In particular, removing `fallback={<Scene/>}` then nothing was rendered i.e. render did not run in worker (or maybe I couldn't see it). Couldn't see anything in React devtools either (you can usually see react-three-fiber trees there).

If you have a working example I can run, I'll try.

**Comment by opeologist** [comment]
> Hey [@opeologist](https://github.com/opeologist) just had a go, but couldn't get the basics working.
> 
> In particular, removing `fallback={<Scene/>}` then nothing was rendered i.e. render did not run in worker (or maybe I couldn't see it). Couldn't see anything in React devtools either (you can usually see react-three-fiber trees there).
> 
> If you have a working example I can run, I'll try.

yeah, we had to pull in the code locally, because it seems like there was a bug that isn't released in the version on npm. you need to copy the current Canvas component locally to get it to work until they release a new version with the fixed code

**Comment by rob-myers** [comment]
@opeologist how do I use it?
So far I've:
```sh
npm i @react-three/offscreen@pmndrs/react-three-offscreen#main
```
but seems I need to build.

**Comment by opeologist** [comment]
> [@opeologist](https://github.com/opeologist) how do I use it? So far I've:
> 
> npm i @react-three/offscreen@pmndrs/react-three-offscreen#main
> but seems I need to build.

i literally copied [Canvas.tsx](https://github.com/pmndrs/react-three-offscreen/blob/main/src/Canvas.tsx) into my local project and imported that component over using the npm's canvas element

**Comment by rob-myers** [comment]
Sorry @opeologist still no luck.
If you can provide working version I will try to fiddle.
However, it's unclear to me if hot-module-reloading will work properly in a web-worker-rendered component.

If you only want to reset your app onchange code, you could create/terminate the worker in a `React.useEffect` in App component.

**Comment by opeologist** [comment]
yeah, no worries, i think we're gonna just have to use the fallback during development to get the HMR DX... ty for trying to help out, though! maybe once this repo's revived something can be spoken of related to hot module reloading when rendering through the offscreencanvas worker 🤷
